sympify : Added support for translating basic numpy datatypes

ArifAhmed1995 opened the pull request:
This is reference to [#11086](https://github.com/sympy/sympy/issues/11086)

Previously for 64-bit systems running Python 3.x numpy's int64 was converted into float.This fix converts it to 'int'.

UPDATE : sympify translators for basic numpy datatypes have been added.Unit tests have also been written.

Previously :

```
import numpy as np
import sympy as sp

x = sp.symbols('x')
print(x * 1)
print(x * np.int64(1))

```

Output : 

```
x
1.0*x

```

```
With the fix : 
import numpy as np
import sympy as sp

x = sp.symbols('x')
print(x * 1)
print(x * np.int64(1))
```

Output : 

```
x
x
```

Ref : http://stackoverflow.com/questions/24914735/convert-numpy-list-or-float-to-string-in-python

ArifAhmed1995 commented:
@moorepants Can you review this now ? I've added unit tests too.

Ref : [11595](https://github.com/sympy/sympy/pull/11595)

asmeurer commented on sympy/core/sympify.py:
```diff
@@ -266,6 +266,17 @@ def sympify(a, locals=None, convert_xor=True, strict=False, rational=False,
+    #Support for basic numpy datatypes.
+    if type(a).__module__ == 'numpy':
+        import numpy
+        if(numpy.isscalar(a)):
```
Don't use parentheses with if statements.

asmeurer commented on sympy/core/sympify.py:
```diff
@@ -266,6 +266,17 @@ def sympify(a, locals=None, convert_xor=True, strict=False, rational=False,
+            try:
+                return sympify(numpy.asscalar(a))
+            except RuntimeError:
+                return sympify(numpy.float64(a))
```
But SymPy has `Float`, which is arbitrary precision. We should probably create a Float with matching precision as the numpy float.

asmeurer commented on sympy/core/tests/test_sympify.py:
```diff
@@ -511,6 +511,37 @@ def test_sympify_set():
 
+def test_numpy():
+    try:
+        import numpy as np
```
You should use sympy.external.import_module here (look how it's done in the lambdify tests).

asmeurer commented on sympy/core/sympify.py:
```diff
@@ -266,6 +266,21 @@ def sympify(a, locals=None, convert_xor=True, strict=False, rational=False,
+                from sympy.core import Float
+                a = str(list(np.reshape(np.asarray(a),
+                                (1, np.size(a)))[0]))[1:-1]
+                return Float(a)
```
Does this use the correct precision? Do you need to pass in the precision as the second argument?

ArifAhmed1995 replied:
The numpy float is converted to it's string representation, and then to sympy Float, hence no precision is lost.The precision is exactly same as the input numpy float.
At first I tried using a = str(a) , but that led to precision loss
`k = np.float128(1.1234567891234568)` ----> `str(k)` ---> `'1.12345678912'`.

With the above translation : 
```
>>> k = np.float128(1.1234567891234568)
>>> k
1.1234567891234568116
>>> sympify(k)
1.1234567891234568116
>>> sympify(k) == k
True
```

ArifAhmed1995 replied:
@asmeurer Should I implement this in another way ?

asmeurer replied:
If it uses the correct precision for the various numpy float types then this is fine.

ArifAhmed1995 replied:
@asmeurer Can you review this again ? I've made the equality tests better.

ArifAhmed1995 committed:
Use chop=True to discard insignificant terms

ArifAhmed1995 committed:
Make corrections and rewrite tests

Prvent mixup due to another unrelated commit

Use sympy.external.import_module to import numpy

Revert deleted space

Make corrections and rewrite tests.

ArifAhmed1995 committed:
Fix whitespace,convert only if numpy scalar.

moorepants commented on sympy/core/tests/test_sympify.py:
```diff
@@ -512,6 +513,48 @@ def test_sympify_set():
+        assert sympify(
+            np.float64(1.1234567891234)) == np.float64(1.1234567891234)
+        assert sympify(np.complex64(1 + 2j)) == 1.0 + 2.0*I
+        assert sympify(np.complex128(1 + 2j)) == 1.0 + 2.0*I
```
Shouldn't these assertions be checking if they are properly converted to sympy types?

ArifAhmed1995 replied:
Done.

ArifAhmed1995 committed:
Better equality tests

asmeurer commented on sympy/core/sympify.py:
```diff
@@ -252,6 +252,22 @@ def sympify(a, locals=None, convert_xor=True, strict=False, rational=False,
+                                            (1, np.size(a)))[0]))[1:-1]
+                    return Float(a)
+                except NotImplementedError:
+                    print('Translation for numpy float : %s'
```
Don't put print statements in code. The correct error here is `SympifyError`.

asmeurer commented:
Does this maintain the appropriate precision for each float type? Can you add tests for that?

asmeurer commented on sympy/core/tests/test_sympify.py:
```diff
@@ -512,6 +513,50 @@ def test_sympify_set():
+        except AttributeError:  #float128 does not exist on all platforms
+            pass
+    else:
+        skip('numpy not installed.Abort numpy tests.')
```
Missing space after .

asmeurer replied:
You can just put 

```py
if not np:
    skip('numpy not installed')
```

at the top and avoid indenting the whole test.

ArifAhmed1995 committed:
Implement requested changes.

ArifAhmed1995 commented on sympy/core/tests/test_sympify.py:
```diff
@@ -512,6 +513,51 @@ def test_sympify_set():
+        sympify(np.uint64(1234567891234567891)), S(1234567891234567891))
+    assert equal(sympify(np.float16(1.123)), S(1.123))
+    assert equal(sympify(np.float32(1.1234567)), S(1.1234567))
+    assert equal(sympify(np.float64(1.1234567891234)), S(1.1234567891234))
```
@asmeurer 
I've maxed out the input in these tests. `numpy` gives imprecise results beyond these.
```
>>> np.float64(1.1234567891234)
1.1234567891234
>>> np.float64(1.12345678912345)
1.1234567891234499
```
Regardless of numpy output, calling sympify on it results in the exact numpy result (i.e. numerically),  but as a sympy number.
I guess appropriate precision is maintained.

asmeurer replied:
I don't think any precision is lost, since the default Float precision corresponds (roughly) to a 64-bit machine float (53 bits, which you can see from the `_prec` attribute of Float). But it would seem to me that perhaps the lower floats should create a float with less precision. For instance, a 32-bit float only has 24 bits of precision (7 decimal digits). See https://en.wikipedia.org/wiki/Floating_point#Internal_representation.

Float takes a decimal precision, which isn't exact. It may be better to convert the float to the appropriate `_mpf_` tuple. Even so, the Float constructor ignores the fourth field in a tuple over the decimal prec. It would be nice to either change that (I think I tried once but got some test failures), or add support to pass in binary precision directly to the Float constructor.

ArifAhmed1995 replied:
I'm trying to use `numpy.finfo(numpy.float).precision` to supply required no. of max digits to the Float constructor. However, it seems to be inadequate. For example `numpy.float16` can handle max no. of 4 decimal digits, but precision as reported by `numpy.finfo` is 3. It's also 1 or 2 units lower for the other float types as well.

asmeurer replied:
I think it will be necessary to use the binary precision.

ArifAhmed1995 committed:
Create Floats according to required precision

ArifAhmed1995 commented on sympy/core/sympify.py:
```diff
@@ -252,6 +252,24 @@ def sympify(a, locals=None, convert_xor=True, strict=False, rational=False,
+                    prec = np.finfo(a).nmant
+                    a = str(list(np.reshape(np.asarray(a),
+                                            (1, np.size(a)))[0]))[1:-1]
+                    return N(a, repr_dps(prec)-1)
```
@asmeurer This uses the precision as required by different types of floats.

ArifAhmed1995 replied:
@asmeurer Can you please review again ?

asmeurer replied:
This looks good. It's probably better to use `Float` instead of `N` here. `N` calls `sympify`, which creates a Float, then evalf, which creates another Float. So it's better to use `Float` here and just create one Float. 

I wish the Float constructor had support for setting the binary precision directly, instead of just the base-10 precision, but that is a separate issue.

asmeurer commented:
Just the one comment. Looks good otherwise.

ArifAhmed1995 committed:
Merge branch 'master' into core

ArifAhmed1995 committed:
Pass binary precision to Float

ArifAhmed1995 commented on sympy/core/sympify.py:
```diff
@@ -256,6 +256,23 @@ def sympify(a, locals=None, convert_xor=True, strict=False, rational=False,
+                    prec = np.finfo(a).nmant
+                    a = str(list(np.reshape(np.asarray(a),
+                                            (1, np.size(a)))[0]))[1:-1]
+                    return Float(a, precision=prec)
```
@asmeurer Now the Float constructor accepts binary precision. [#12227](https://github.com/sympy/sympy/pull/12227)
